clippy_lints/src/ifs_as_logical_ops.rs

```diff
@@ -0,0 +1,109 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::higher::If;
+use clippy_utils::source::{SpanRangeExt, walk_span_to_context};
+use clippy_utils::sugg::Sugg;
+use clippy_utils::{is_else_clause, peel_blocks};
+use rustc_ast::LitKind;
+use rustc_errors::Applicability;
+use rustc_hir::{Expr, ExprKind};
+use rustc_lint::{LateContext, LateLintPass, LintContext};
+use rustc_session::declare_lint_pass;
+
+declare_clippy_lint! {
+    /// ### What it does
+    ///
+    /// Warn about cases where `x && y` could be used in place of an if condition.
+    ///
+    /// ### Why is this bad?
+    ///
+    /// `x && y` is more standard as a construction, and makes it clearer that this is just an and.
+    /// It is also less verbose.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// # fn a() -> bool { false }
+    /// fn b(b1: bool) -> bool {
+    ///  if b1 { a() } else { false}
+    /// }
+    ///
+    /// ```
+    ///
+    /// Could be written
+    ///
+    /// ```no_run
+    /// # fn a() -> bool { false }
+    /// fn b(b1: bool) -> bool {
+    ///   b1 && a()
+    /// }
+    /// ```
+    #[clippy::version = "1.89.0"]
+    pub IFS_AS_LOGICAL_OPS,
+    nursery,
+    "`if` conditions that can be rewritten as logical operators"
+}
+declare_lint_pass!(IfsAsLogicalOps => [IFS_AS_LOGICAL_OPS]);
+
+impl<'tcx> LateLintPass<'tcx> for IfsAsLogicalOps {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, e: &'tcx Expr<'tcx>) {
+        // Make sure the if block is not an if-let block.
+        if let Some(If {cond: if_cond, then: cond_inner, r#else: Some(els)}) = If::hir(e)
+            && let ExprKind::Block(if_block, _label) = cond_inner.kind
+            // Check if the if-block has only a trailing expression
+            && if_block.stmts.is_empty()
+            && let Some(if_block_inner_expr) = if_block.expr
```

Comment:
The first block of an if is often called `then`, I think that's a pretty good convention
```suggestion
        if let Some(If {cond: if_cond, then, r#else: Some(els)}) = If::hir(e)
            && let ExprKind::Block(then_block, _label) = then.kind
            // Check if the then-block has only a trailing expression
            && then_block.stmts.is_empty()
            && let Some(then_block_inner_expr) = then_block.expr
```